cpp/src/groupby/sort/group_rank_scan.cu

```diff
@@ -0,0 +1,120 @@
+    thrust::tabulate(rmm::exec_policy(stream),
+                     mutable_ranks.begin<size_type>(),
+                     mutable_ranks.end<size_type>(),
+                     [comparator,
+                      d_col_order = *d_col_order,
+                      labels      = group_labels.data(),
+                      offsets     = group_offsets.data()] __device__(size_type row_index) {
+                       auto group_start = offsets[labels[row_index]];
+                       if (row_index == group_start) { return 1; }
+                       bool const lhs_is_null{d_col_order.is_null(row_index)};
+                       bool const rhs_is_null{d_col_order.is_null(row_index - 1)};
+                       if (lhs_is_null && rhs_is_null) {
+                         return 0;
+                       } else if (lhs_is_null != rhs_is_null) {
+                         return row_index - group_start + 1;
+                       }
+                       return comparator(row_index, row_index - 1) ? 0
+                                                                   : row_index - group_start + 1;
+                     });
+  } else {
+    thrust::tabulate(rmm::exec_policy(stream),
+                     mutable_ranks.begin<size_type>(),
+                     mutable_ranks.end<size_type>(),
+                     [comparator,
+                      labels  = group_labels.data(),
+                      offsets = group_offsets.data()] __device__(size_type row_index) {
+                       auto group_start = offsets[labels[row_index]];
+                       return row_index != group_start && comparator(row_index, row_index - 1)
+                                ? 0
+                                : row_index - group_start + 1;
+                     });
+  }
+
+  thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                group_labels.begin(),
+                                group_labels.end(),
+                                mutable_ranks.begin<size_type>(),
+                                mutable_ranks.begin<size_type>(),
+                                thrust::equal_to<size_type>{},
+                                DeviceMax{});
+
+  return ranks;
+}
+}  // namespace
+std::unique_ptr<column> rank_scan(column_view const& order_by,
+                                  device_span<size_type const> group_labels,
+                                  device_span<size_type const> group_offsets,
+                                  rmm::cuda_stream_view stream,
+                                  rmm::mr::device_memory_resource* mr)
+{
+  if (order_by.type().id() == type_id::STRUCT) {
+    bool nested_nulls =
+      std::any_of(order_by.child_begin(), order_by.child_end(), [](auto const& col) {
+        return has_nested_nulls(table_view{{col}});
+      });
+    if (nested_nulls) {
+      return generate_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
+    }
```

Comment:
Similar to above, just use this if we can:
```
if ((order_by.type().id() == type_id::STRUCT && has_nested_nulls(table_view{{order_by}})) 
      || order_by.has_nulls()) {
      return generate_dense_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
} 
return generate_dense_ranks<false>(order_by, group_labels, group_offsets, stream, mr);
```

Reply:
```
  if ((order_by.type().id() == type_id::STRUCT && has_nested_nulls(table_view{std::vector<column_view>{order_by.child_begin(), order_by.child_end()}}))
      || (order_by.type().id() != type_id::STRUCT && order_by.has_nulls())) {
      return generate_dense_ranks<true>(order_by, group_labels, group_offsets, stream, mr);
} 
return generate_dense_ranks<false>(order_by, group_labels, group_offsets, stream, mr);
```